File: java/src/main/java/Perceptron.java
public class Perceptron {
    double errorThreshold;
    double weights[];
    int trainData[][];
    int target[];

    public void toTrain(int _trainData[][], int _target[]) {
        this.toTrain(_trainData, _target, 0.1);
    }

    public void toTrain(int _trainData[][], int _target[], double _errorThreshold) {
        errorThreshold = _errorThreshold;
        trainData = _trainData;
        target = _target;

        weights = new double[_trainData[0].length + 1];
        for(int i=0; i < weights.length; i++)
            weights[i] = 1;

        boolean isItOver = true;
        do{
            isItOver = true;
            for (int line=0; line< target.length ; line++) {
                int error = calculateError(line);
                if (error != 0) {
                    calculateWeights(error, line);
                    isItOver = false;
                }
            }
        }while(isItOver == false);
    }

    public void calculateWeights(int error, int line) {
        weights[0] = weights[0] + errorThreshold * error;
        for (int i=1; i < weights.length ; i++) {
            weights[i] = weights[i] + errorThreshold * error * trainData[line][i-1];
        }
    }

    public int calculateError(int line) {
        return target[line] - calculateOutput(trainData[line]);
    }

    public int calculateOutput(int data[]) {
        double total = weights[0];
        for (int i=1; i < weights.length ; i++) {
            total += weights[i] * data[i-1];
        }
        return (total >=0) ? 1 : 0;
    }
}

File: go/perceptron.go
package perceptron

var Weights []float64
var TrainData [][]int
var Target []int
var ErrorThreshold float64

func ToTrain(trainData [][]int, target []int) {
	ToTrainAccurracy(trainData, target, 0.1)
}

func ToTrainAccurracy(trainData [][]int, target []int, errorThreshold float64) {
	TrainData = trainData
	Target = target
	ErrorThreshold = errorThreshold
	Weights = []float64{1}
	for i := 0; i < len(TrainData[0]); i++ {
		Weights = append(Weights, 1)
	}

	for isItOver := false; !isItOver; {
		isItOver = true
		for line, _ := range Target {
			error := CalculateError(line)
			if error != 0 {
				CalculateWeights(error, line)
				isItOver = false
			}
		}
	}
}

func CalculateOutput(data []int) int {
	var total float64 = 0
	for index, value := range append([]int{1},data...) {
		total += float64(value) * Weights[index]
	}
	if total >= 0 {return 1} else { return 0}
}

func CalculateError(line int) int {
	return Target[line] - CalculateOutput(TrainData[line])
}

func CalculateWeights(error int, line int) {
	for index, value := range append([]int{1},TrainData[line]...) {
		Weights[index] = Weights[index] + ErrorThreshold * float64(error) * float64(value)
	}
}